Bug: AUC start 0 (which also needs Feature: C0 imputation)

Gero1999 opened the pull request:
## Issue
Closes #2 #60

## Description
After running NCA, AUC calculations always consider the discrete start (start at where first concentration sample was collected). This can indeed alter slightly the real values of the AUC. This is a consequence of 2 problems:

1) Doses times are being considered at the first measure, instead of at the real dose time
2) No imputed measure at dose time is consdiered

Change description.

## Definition of Done
C0 imputation option for the user specified in Setup > Data Selection

## How to test
- [x] At "NCA">"NCA Results", when C0 is imputed it is expected that start=AFRLT-ARRLT (being AFRLT-ARRLT= 0 in DOSNO=1)
- [x] At "NCA">"NCA Results", when C0 is not imputed it is expected that start={time of 1st observation} 
- [x] AUC values  (auclast, aucinf...) from start=AFRLT-ARRLT are calculated when C0 is imputed
- [x] All variables that should only be based on observed data (cmax, tmax...) are not considering C0 imputations
- [x] Test that PKNCA also deals well with predoses as duplicates for the NCA
- [x] Check slope selector plots and NCA results remain well after running NCA
- [x] Check if in general C0 imputation logic works well (NCA experts): 
if `C0 exists` ~ `C0=C0`
if `DOSNO=1 & IV bolus (ADOSEDUR=0, ROUTE=intravascular)` ~ `C0=0`
if `DOSNO>1 & not IV bolus` ~ `C0=predose`
if `IV bolus and monoexponential data` ~ `logslope`
if `IV bolus and not monoexponential data` ~ `C0=C1`

## Contributor checklist
- [x] Code passes lintr checks
- [x] Code passes all unit tests
- [x] New logic covered by unit tests
- [x] New logic is documented

## Notes to reviewer

Anything that the reviewer should know before tacking the pull request?

js3110 reviewed (changes requested):
Nice work! Some requests:

Fixes:
- [ ] App crashes if AUC rule sets are selected
- [ ] Slope selector not working- i tested an exclusion and the app crashed. Haven't tried selections.
- [ ] Data table in Data selection tab is not sorted- needs to be sorted by USUBJID for better visualisation.

Enhancements:

- It would be good if the user can see an updated data table after the imputations, so they know what the data being used looks like. If this is too complex then maybe you can add an explanation of what the imputation is doing. But it would be good if they can see what data is being created.
- Some clarification/relabelling is required in terms of what the imputation does. 'C0 missing' is not always accurate, for example if the predose value is present in the data set but ATime is <0, it gets excluded, but its not technically missing. Some more clarification on this would be good for the user so they know that imputation is also for predose samples.
- Predose samples/ duplicates are not being included. Need to figure out a way to ensure they are- by creating duplicates manually or making sure PKNCA includes them.

m-kolomanski reviewed (changes requested):
Hey,

Are the diffs displayed by github accurate? Seems like a lengthy change, are there any artifacts/errors that happened due to merge (it did happen to me on occasions)? If the diffs are inaccurate, please point out the actual changes and disregard the below.

I do not have the time to do a detailed review (it is quite a lengthy change), but after a quick look there is some general things I want to point out:

1. Please do not any files that solely generate UI to the `/R` folder - this is a folder for our logic. File `partial_auc_input` belongs in the `inst/shiny` folder, either in `/modules` (if there is any server logic attached, here it isnt, but might be added) or folder like `functions`. We can also create a new folder for UI elements, called `ui` or `elements` or `widgets` or whatever you find suitable.
2. If you are creating new files or adding new functions in the `/R` folder, please make sure that this logic is testable and do take some effort to write at least basic unit tests. Adding tests back to existing logic is tough and time consuming, so we should at least make sure that our testing suite keeps up with any new features or major changes to the old ones.
3.

Gero1999 commented:
> Hey,
> 
> Are the diffs displayed by github accurate? Seems like a lengthy change, are there any artifacts/errors that happened due to merge (it did happen to me on occasions)? If the diffs are inaccurate, please point out the actual changes and disregard the below.
> 
> I do not have the time to do a detailed review (it is quite a lengthy change), but after a quick look there is some general things I want to point out:
> 
> 1. Please do not any files that solely generate UI to the `/R` folder - this is a folder for our logic. File `partial_auc_input` belongs in the `inst/shiny` folder, either in `/modules` (if there is any server logic attached, here it isnt, but might be added) or folder like `functions`. We can also create a new folder for UI elements, called `ui` or `elements` or `widgets` or whatever you find suitable.
> 2. If you are creating new files or adding new functions in the `/R` folder, please make sure that this logic is testable and do take some effort to write at least basic unit tests. Adding tests back to existing logic is tough and time consuming, so we should at least make sure that our testing suite keeps up with any new features or major changes to the old ones.

Hello and happy new year @m-kolomanski !

You are right, this is a very lengthy change but I re checked and everything was intentional. Most of them are new created manuals for each function. As requested I also:

1. Deleted the extra UI function from `\R`
2. Added tests for all functions called in ``format_data,R`, `create_start_impute` and `PKNCA_impute_method_additions`

Don't hesitate to let me know if there is anything else I can do or if there are other concerns that I can address! I also asked @js3110 for re-review to make sure all looks as expected.

js3110 reviewed (approved):
LGTM

Need to create a new issue for the auc calculation to avoid using cmax automatically.

m-kolomanski commented on tests/testthat/test-format_data.R:
```diff
@@ -0,0 +1,326 @@
+  expect_equal(df_conc$IX, rep(1:5, 2))
+
+  # Test if df_conc can be used with PKNCAconc by testing its output
+  expect_s3_class(
+    PKNCAconc(
+      df_conc,
+      formula = AVAL ~ TIME | STUDYID + PCSPEC + DRUG + USUBJID / ANALYTE,
+      exclude_half.life = "exclude_half.life",
+      time.nominal = "NFRLT"
+    ),
+    "PKNCAconc"
+  )
```
issue: In this case, if `df_conc` was somehow wrong, we would expect an error from `PKNCAconc` function, so I would wrap that in `expect_no_error` to know exactly if a) an error occured or b) `PKNCAconc` returned something strange.
```suggestion
  expect_no_error({
    expect_s3_class(
      PKNCAconc(
        df_conc,
        formula = AVAL ~ TIME | STUDYID + PCSPEC + DRUG + USUBJID / ANALYTE,
        exclude_half.life = "exclude_half.life",
        time.nominal = "NFRLT"
      ),
      "PKNCAconc"
    )
  })
```

m-kolomanski commented on tests/testthat/test-format_data.R:
```diff
@@ -0,0 +1,326 @@
+               rep(1:5, 4))
+
+  # Test if df_conc can be used with PKNCAconc by testing its output
+  expect_s3_class(
```
issue: As above, I would wrap that in `expect_no_error`.

m-kolomanski commented on tests/testthat/test-format_data.R:
```diff
@@ -0,0 +1,326 @@
+                    "TIME", "IX", "ROUTE") %in% colnames(df_dose)))
+
+  # Test if df_dose can be used with PKNCAdose by testing its output
+  expect_s3_class(
```
issue: As above, I would wrap that in `expect_no_error`.

m-kolomanski commented on tests/testthat/test-format_data.R:
```diff
@@ -0,0 +1,326 @@
+  expect_true(all(myintervals$type_interval == "main"))
+
+  # Test if myintervals can be used with PKNCAdata by testing its output
+  expect_s3_class(
```
issue: As above, I would wrap that in `expect_no_error`.

m-kolomanski commented on R/format_data.R:
```diff
@@ -1,121 +1,162 @@
+    stop(paste("Missing required columns:", paste(missing_columns, collapse = ", ")))
+  }
+
+  data <- ADNCA %>%
```
issue: variable assignment not needed, can just return the pipe result
```suggestion
  ADNCA %>%
```

m-kolomanski reviewed (changes requested):
Hey, happy new year to you too and thank you for all the hard work!

I have to request some changes and have some suggestions/nitpicks to consider. I know that some of my comments refer to lines that were not changed in this PR, but since we are working on those functions I think it would be beneficial to go back and improve the code, even if it is not directly related to the changes.

Gero1999 commented on R/PKNCA_impute_method_additions.R:
```diff
@@ -0,0 +1,69 @@
+PKNCA_impute_method_start_log <- function(conc, time, start, end, ..., options = list()) { # nolint
+
+  ret <- data.frame(conc = conc, time = time)
+  mask_start <- time %in% start
```
I think you are right. It is an unnecessary prevention of NAs in case `time` contains them, but it should not at this point

Gero1999 commented on R/PKNCA_impute_method_additions.R:
```diff
@@ -0,0 +1,69 @@
+
+PKNCA_impute_method_start_log <- function(conc, time, start, end, ..., options = list()) { # nolint
+
+  ret <- data.frame(conc = conc, time = time)
```
I asked Bill because I actually followed his syntax from the other parallel functions. Apparently is from the word `return` (of the function). But I will change it to something more intuitive now

Gero1999 replied:
I put `d_conc_time`, hope that makes more sense!

Gero1999 commented on R/create_start_impute.R:
```diff
@@ -0,0 +1,102 @@
+#' \dontrun{
+#' mydata <- create_start_impute(mydata)
+#' }
+create_start_impute <- function(mydata) {
```
I agree with this but I would prefer to change all `mydata` calls together at once to something we set. Maybe in an independent issue for renaming variables would be the best

Gero1999 commented:
Thanks for your reviews and feedback, I hope that now all looks fine! @js3110 @m-kolomanski

m-kolomanski reviewed (changes requested):
Hey, looks much better, thank you for all the hard work! I still do have some feedback, most of it regards style and consistency rather thank quality of the code, but also there seems like some of my comments from the previous review were not addressed?

js3110 commented:
@Gero1999 it looks good and thank you for all the hard work on this! Unfortunately I have noticed a couple of things:

1. In the data selection, after you select analyte and dose number, the table shown does not represent the selections, as I can still see all the dose numbers. We want this table for the user to see exactly the data that will be used for NCA, so either we need to filter the dose number here, or move the dose number selectInput. I will let you decide.
2. For the dummy data, Doseno >3 does not work and the app crashes. I get the following error: Error in if: argument is of length zero. I double checked and I do not get this issue in main, so it is related to something in this branch. Also, for DOSNO = 2, USUBJID 11102 has empty cells for lambda z calculation, but no explanation why and because of this the slope selection crashes. This could be created as a separate bug issue to fix later or you can see if its something to fix now.

Let me know if you have any questions or want any support :)

m-kolomanski commented on R/PKNCA_impute_method_additions.R:
```diff
@@ -0,0 +1,75 @@
+
+  d_conc_time <- data.frame(conc = conc, time = time)
+  if (!any(time == start)) {
+    all_concs <- conc[time >= start  &  time <= end]
+    all_times <- time[time >= start  &  time <= end]
```
nitpick, style: 
```suggestion
    all_concs <- conc[time >= start & time <= end]
    all_times <- time[time >= start & time <= end]
```

m-kolomanski commented on R/create_start_impute.R:
```diff
@@ -0,0 +1,101 @@
+    ungroup() %>%
+    rowwise() %>%
+    mutate(
+      impute = case_when(
+        # Start concentration already present: No imputation (NA)
+        !!sym(time_column) == start & !is.na(!!sym(conc_column)) ~ NA,
+
+        # 1st dose with not IV bolus or analyte =! drug : Start concentration is 0
+        is.first.dose & (!is.ivbolus | !is.analyte.drug) ~ "start_conc0",
+
+        # Posterior doses not IV bolus or analyte =! drug : Start concentration shifts to predose
+        !is.first.dose & (!is.ivbolus | !is.analyte.drug) ~ "start_predose",
+
+        # IV bolus with analyte = drug : Start concentration is log-backextrapolated (if possible)
+        is.ivbolus & is.analyte.drug & is.possible.c0.logslope ~ "start_logslope",
+
+        # IV bolus with analyte = drug and not possible logslope: Start concentration is 1st conc
+        is.ivbolus & is.analyte.drug ~ "start_c1"
```
Nice, looks clean, thank you for additional comments with explanations!

m-kolomanski commented on inst/WORDLIST.txt:
issue(non-blocking): Please remove this file, it is redundant.

m-kolomanski reviewed (approved):
From my side, looks good! Thank you for all the hard work, this is a lengthy change, sure it was not easy!

I am approving, but before merge, please remove the `inst/WORDLIST.txt` file, as it is redundant - the spelllcheck only uses `WORDLIST` file.

js3110 reviewed (approved):
LGTM !!
